### ISSUE TITLE
Encryption minimum Android version - "Could not encrypt file content. The file was not saved"

### ISSUE THREAD

**Issue by AnFlare** [gsantner/markor#1173]
#### General information

* **App version:** 2.5.0 F-Droid
* **System:** Android 7.0

#### Description
The issue itself is nicely visible on the screenshot below
_(sorry, IDK how to hide it to spoiler or something else)_

#### How to reproduce:
1. go to Settings, General, then enable Experimental features there
2. go to notes' list, press "+" floating button
3. set any filename, Type: Markdown, Template: Empty file, checkbox "encrypt file content" enabled
4. write anything in this note
5. quit note or try to save it by ... button

![Screenshot_20201226-154521](https://user-images.githubusercontent.com/7705012/103150261-397a9980-4794-11eb-933b-e3307a8b442e.png)

#### Logcat
[logreader_2020-12-26_16-00.log](https://github.com/gsantner/markor/files/5742983/logreader_2020-12-26_16-00.log)

**Comment by gsantner** [comment]
Have you used the file extension with ".jenc" which the checkbox adds? This is mandatory here. If filename not ends with .jenc add it on the end and reopen.

**Comment by AnFlare** [comment]
@gsantner 
1. Huh, sorry, forgot to mention it  🙈  It was definitely set. ("123", just for testing)
2. Yes, the app adds this extension automatically and I didn't delete it. Just tried to write there and save.

**Comment by gsantner** [comment]
OK hm works all fine for me. @opensource21 do you have an idea?

**Comment by gsantner** [comment]
@AnFlare 
If you create a new encrypted file now and write something in this new one...does it work now? Or has it happened only once for that first file?

**Comment by opensource21** [comment]
I never installed the app via f-droid. I'm surprised to see no error in the log. Only
`type=1400 audit(0.0:6116): avc: denied { read } for comm="gsantner.markor" name="madev0" dev="tmpfs" ino=7280 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:microarray_finger_device:s0 tclass=chr_file permissive=0`
Doesn't look good. 
@AnFlare Can you provide more information about your android, specially which manufactor it builds. Have you tested the google-play version?

**Comment by gsantner** [comment]
google play, fdroid version or debug version ... doesn't matter, theres no major or relevant difference.

**Comment by AnFlare** [comment]
@gsantner @opensource21 
> If you create a new encrypted file now and write something in this new one...does it work now? Or has it happened only once for that first file?

Nope. I tried to create encrypted file anew, I tried Google Play version 2.5.0. No way. In any case there's only "Could not encrypt file content"
As for manufacturer, it's one of those chinese companies (Doogee X60L, to be more exact). Though I'm surprised about this issue, because I never had any of them while using this phone, almost 3 years so far.

**Comment by opensource21** [comment]
I take look into the code. The message was only used at `net.gsantner.markor.util.DocumentIO#writeContent`. But there I found
```
} catch (JavaPasswordbasedCryption.EncryptionFailedException e) {
    Log.e(DocumentIO.class.getName(), "loadDocument:  enrypt failed for File " +
            document.getFile().getAbsolutePath() + ". " + e.getMessage(), e);
    Toast.makeText(context, R.string.could_not_encrypt_file_content_the_file_was_not_saved, Toast.LENGTH_LONG).show();
    ret = false;
}
```

If we can get the log and the exception I could analyze it. Without this I can only guess what the reasons could be. 

@gsantner  Do you have any idea how @AnFlare could provide us the complete log?

**Comment by AnFlare** [comment]
@opensource21 
> how @AnFlare could provide us the complete log?

I tried it with another logcat reader. May be it will be more helpful then the first one. Here's the full log while trying to encrypt:
[Full.txt](https://github.com/gsantner/markor/files/5745429/Full.txt)

I made a search there by "encr" and then found this:
https://pastebin.com/gA3BKnFp
`12-28 04:38:23.889 E/net.gsantner.markor.util.DocumentIO(24300): Caused by: other.de.stanetz.jpencconverter.JavaPasswordbasedCryption$EncryptionFailedException: Error creating key from password: PBKDF2WithHmacSHA512 SecretKeyFactory not available`

And this at last:
https://stackoverflow.com/a/60150663
Feels like the encryption in its current implementation can't work with Android 7 and below.
But it's nothing more than my hypothesis.

**Comment by gsantner** [comment]
I have something in memory that it was part of some PR discussion - that it will only work starting from Android version XY. Might be just not noted.

**Comment by opensource21** [comment]
Well this was what I expected the encryption algorithm wasn't present on the device. This is the pain with the reflection. It's hard to find there a safe solution.

@gsantner The encryption only works for Android KitKat (this we discussed at the PR). This problem was solved by checking the version via if statements. We now have 3 possibilities:

1. We disabled encryption on devices older than API 26 Android 8  (20% of the market).
2. We try out which versions are supported by Android 5 and I created a new version which choose these weaker encryption method. Then we choose the right encryption method via checking the version. The main drawback is that if someone uses the desktop app and create a file with Version1 then it can't be decrypted at the smartphone and we will run into an error. Furthermore AES 128 isn't really safe.
3. Perhaps bouncy-castle would be an option, but then we included a big jar for this option.

**Comment by opensource21** [comment]
@AnFlare Thanks for the information and the analysis you did.

**Comment by gsantner** [comment]
we require android 8 for encryption, thats it then. I dont wanna experiment with the algorithms - it potentially destroys everything encryped that is currently there on working devices.

**Comment by AnFlare** [comment]
@opensource21 
> The encryption only works for Android KitKat

So it worked before?

> We now have 3 possibilities

Why you don't consider "other" way, written on Stackoverflow? There's something wrong with it?

> The main drawback is that if someone uses the desktop app

Why not store encryption mode just inside the note itself? Why not to implement a toggle in settings to choose needed encryption mode? (with brief explanation of possible drawbacks, may be)

@gsantner 
> we require android 8 for encryption, thats it then

I hope so much there's a workaround to solve this case. I have no possibility to upgrade my device in any foreseeable future and I live in third world country with bloody, despotic, totalitarian & criminal government.  Tried every notes encryption app on the internet and all of them either can't do text emphasizing (bold, italic, etc.) or updated hundreds years ago, or other. Markor was my only hope.

**Comment by gsantner** [comment]
the highest priority is that nothing breaks for people that **already have encrypted notes**, no matter what happens.

**Comment by opensource21** [comment]
@AnFlare 
> So it worked before?

No. I want to say: Lower than Kitkat you can't store the password, so there is definitely no encryption.
> Why you don't consider "other" way, written on Stackoverflow? There's something wrong with it?

The other way is still the same problem - it's just code with weak encryption.
>Why not store encryption mode just inside the note itself? Why not to implement a toggle in settings to choose needed encryption mode? (with brief explanation of possible drawbacks, may be)

The encryption mode is stored in the file. That's the reason why it's possible to change the version and provide different encryption mode. But if you create at desktop a version with strong encryption it can't be decrypted on old devices.

Changing something at encryption is always with some risk. And it's really cumbersome to test it, so I can understand gsantner. 
He is the one who makes the decision.

You can try https://turtlapp.com/ I hope it works on your device. Please give it a try and check it. I use it for years. It's only a pain if the server is down.

Other candidates could be
- https://standardnotes.org/
- https://laverna.cc/

Please try them and if nothing work make a note again, perhaps we will find then a solution with markor.

**Comment by gsantner** [comment]
@opensource21 
Is only the (default stored) password the issue, or the encryption itself too?

**Comment by opensource21** [comment]
The password store works, but the derived password can't be build. But even if this would work the encryption will fail. Both is defined by the version-number which is stored in the first bytes of the file. So it's possible to define a weaker encryption with version 2, which will work then with android 7. So I only see problems if an old device try to decrypt and file with was encrypted with version 1.

**Comment by gsantner** [comment]
if something, we add support for minSdk (16) till latest sdk, extending only till Android 7 not makes much benefit.

Though it needs somebody willing to do that, and have enough time for it, at same time while not breaking anything

> Why not store encryption mode just inside the note itself? Why not to implement a toggle in settings to choose needed encryption mode? (with brief explanation of possible drawbacks, may be)

to come back to this question - it's a matter of time resources people are willing to spend for free. While it's only helpful for a minority of users.  
I personal argue, if you have a quite old android version you already have enough other security & privacy issues. Use full device encryption and use private app storage is your best bet what you can do atm at Markor.  
If something bad gets enough access on your device (and i.e. intercept screen or key input), the app encryption wont give you much of a benefit. Given simple things like receiving a image on a messenger (media framework + stagefright) are enough to get root access.

**Comment by opensource21** [comment]
I can reproduce it now and will have a look to support Marshmallow and above.

**Comment by AnFlare** [comment]
@opensource21 
> No. I want to say: Lower than Kitkat you can't store the password, so there is definitely no encryption.

Undestood. That's totally ok.

> Changing something at encryption is always with some risk

I see. If I can be helpful somehow, just let me know. I'm not an Android developer, but my skills are above average, I hope.

>  https://turtlapp.com/
https://standardnotes.org/
https://laverna.cc/

Tried every one of them (and many more on the market). Many of them are very nice, but unfortunately, no one suits my needs. Turtl - doesn't have wysiwyg editor. In StandardNotes I can't try editors extensions before bying. Laverna is the closest to what I need, but while writing something on mobile, it doubles and triples every written word.
As I mentioned before, Markor is the best md files editor with encryption possibility I could find.

There seemed to be 2 other options.
Markor + MiXplorer EncFS container. Or Markor + RCX Crypt remote. But neither of these two are able to cooperate with Markor.

@gsantner 
> I personal argue, if you have a quite old android version you already have enough other security & privacy issues. Use full device encryption and use private app storage is your best bet what you can do atm with current Markor version.

I think every day about it. Full device encryption, I mean. But I fear it will irreversibly slow device down (that's not blazingly fast even in its current state)

**Comment by AnFlare** [comment]
@opensource21 
> I can reproduce it now and will have a look to support Marshmallow and above.

Thank you so much!! I double, if there's anything I can do to help you, just tell me.

**Comment by gsantner** [comment]
Thanks @opensource21 for improving

@AnFlare you may wait for next release, or download test build apk from github actions latest build

**Comment by AnFlare** [comment]
@opensource21 
@gsantner 
Thank you so much!! You're awesome!

As for test build. Do you speak about one from [here](https://github.com/gsantner/markor/actions/runs/450940098)?
It's giving the same error and there seems to be no later builds.

**Comment by gsantner** [comment]
yes from there. If it didnt fix the issue for you, not sure. then we need another fix I guess. please share new logs from that build.

**Comment by AnFlare** [comment]
Here they are.
[logs-2020-12-30-04-27-17.zip](https://github.com/gsantner/markor/files/5752546/logs-2020-12-30-04-27-17.zip)

**Comment by gsantner** [comment]
please be sure to have enabled encryption there and have set password. the test build (Mordor) is complete seperate to Markor.

**Comment by AnFlare** [comment]
Reuploaded logs

**Comment by gsantner** [comment]
```
12-30 04:44:07.874 D/KeymasterHAL(  288): module.cpp:241: Invoke cmdId 18 at vendor/mediatek/proprietary/trustzone/trustkernel/source/external/AndroidServices/keymaster_v2/module.cpp:1636
12-30 04:44:07.881 E/net.gsantner.markor.util.DocumentIO(28472): loadDocument:  enrypt failed for File /storage/emulated/0/Documents/mordor/test1.md.jenc. Can't encrypt text.
12-30 04:44:07.881 E/net.gsantner.markor.util.DocumentIO(28472): other.de.stanetz.jpencconverter.JavaPasswordbasedCryption$EncryptionFailedException: Can't encrypt text.
12-30 04:44:07.881 E/net.gsantner.markor.util.DocumentIO(28472): 	at other.de.stanetz.jpencconverter.JavaPasswordbasedCryption.encryptBytes(JavaPas
.....................
........
other.de.stanetz.jpencconverter.JavaPasswordbasedCryption.encryptBytes(JavaPasswordbasedCryption.java:108)
12-30 04:44:07.881 E/net.gsantner.markor.util.DocumentIO(28472): 	... 30 more
12-30 04:44:07.881 E/net.gsantner.markor.util.DocumentIO(28472): Caused by: java.security.NoSuchAlgorithmException: PBKDF2WithHmacSHA512 SecretKeyFactory not available
12-30 04:44:07.881 E/net.gsantner.markor.util.DocumentIO(28472): 	at javax.crypto.SecretKeyFactory.<init>(SecretKeyFactory.java:249)
12-30 04:44:07.881 E/net.gsantner.markor.util.DocumentIO(28472): 	at javax.crypto.SecretKeyFactory.getInstance(SecretKeyFactory.java:287)
12-30 04:44:07.881 E/net.gsantner.markor.util.DocumentIO(28472): 	at 
```

well your device OS seemingly just not has the cryptos support that it supposed to be for the given Android version. MediaTek, any more words required? 😄

**Comment by opensource21** [comment]
That is strange. Either it's old code level or your phones api level is greater 25 🤨

**Comment by opensource21** [comment]
I only found an error at the logging. DocumentIO.java line 202 it must be
`Log.e(DocumentIO.class.getName(), "writeContent:  enrypt failed for File " +`
instead of
`Log.e(DocumentIO.class.getName(), "loadDocument:  enrypt failed for File " +`
@gsantner can you fix it.

I tested the apk with the emulator API23 Android 6.0.
@AnFlare Can you tested the app on a emulator (need unfortunately about 10 GB Disk-Space). Do you have 7.0 or 7.1?

**Comment by opensource21** [comment]
@gsantner at @AnFlare s phone the wrong version V001 will still chosen even with the new APK. No idea why.

**Comment by opensource21** [comment]
API 24 Android 7.0 on emulator works fine too

**Comment by opensource21** [comment]
Sorry I tested the wrong version. The apk from github is broken. If I compile and install it directly it works.

**Comment by AnFlare** [comment]
@gsantner 
> well your device OS seemingly just not has the cryptos support that it supposed to be there for that Android version.

Is there any way to check it for a certainty?

@opensource21 
> Either it's old code or your phones api level is greater 25
> Do you have 7.0 or 7.1?

That's reported by the system itself (see in logs zip)
version.release: 7.0
version.sdk_int: 24

> Can you tested the app on a emulator

Yup. On the PC? I'll do it a little bit later, definitely.

**Comment by gsantner** [comment]
TL;DR: OK so something might be wrong with the CI's build and/or APK, and the new changes fixed the issue (just you got a bad apk). Correct?

**Comment by opensource21** [comment]
Yes that's correct. 

Am 30. Dezember 2020 13:26:37 MEZ schrieb Gregor Santner <notifications@github.com>:

**Comment by opensource21** [comment]
This one should work

[net.gsantner.markor-v121-2.5.0-flavorAtest-debug.apk.gz](https://github.com/gsantner/markor/files/5755051/net.gsantner.markor-v121-2.5.0-flavorAtest-debug.apk.gz)

**Comment by opensource21** [comment]
I analyzed the apk from https://github.com/gsantner/markor/actions/runs/450940098  if definitly conteins the old version of JavaPasswordbasedCryption.

**Comment by gsantner** [comment]
I checked the CI runs, by configuration newer commits were not build hence they changes didnt get there. Enabled CI again for everything.

**Comment by AnFlare** [comment]
@opensource21 @gsantner 
> This one should work
 
This one works like a charm!! ✨
